lean/commands/research.py

```diff
@@ -157,6 +157,16 @@ def research(project: Path,
 
     container.update_manager().pull_docker_image_if_necessary(research_image, update)
 
+    # Allow the user to map docker ports
+    docker_ports = project_config.get('docker-ports') or {}
+    for outer, inner in docker_ports.items():
+        if outer in run_options["ports"]:
+            raise RuntimeError(f"Port {outer} is already in use, please specify a different host port under 'docker-ports' in your project's config.json")
+        run_options["ports"][outer] = inner
+    
+    # Allow the user to set environment variables
+    run_options["environment"].update(project_config.get('docker-env') or {})
```

Comment:
I recommend moving this logic to [`LeanRunner.get_basic_docker_config()`](https://github.com/QuantConnect/lean-cli/blob/master/lean/components/docker/lean_runner.py#L159), that way these new project configuration properties can be used with `lean backtest`, `lean live`, `lean optimize`, and `lean research` (instead of only `lean research`).